packages/agreement-process/src/services/agreementService.ts

```diff
@@ -1297,3 +1320,32 @@ async function addContractOnFirstActivation(
 
   return agreement;
 }
+
+async function getConsumerFromDelegationOrRequester(
+  eserviceId: EServiceId,
+  delegationId: DelegationId | undefined,
+  organizationId: TenantId,
+  readModelService: ReadModelService
+): Promise<Tenant> {
+  const delegations =
+    await readModelService.getActiveConsumerDelegationsByEserviceId(eserviceId);
+
+  if (delegationId) {
+    // If a delegation has been passed, the consumer is the delegator
+    const delegation = retrieveDelegation(delegations, delegationId);
+
+    assertRequesterIsDelegateConsumer(delegation, eserviceId, organizationId);
+    return retrieveTenant(delegation.delegatorId, readModelService);
+  } else {
+    const hasDelegation = delegations.some(
+      (d) => d.delegateId === organizationId || d.delegatorId === organizationId
+    );
```

Comment:
Shouldn't this check only if the requester is the delegator?
- If there is a delegation and I am the delegator, I cannot create the agreement, so this is correct
- However, there can be a delegation for which I am the delegate: in that case, if the delegationId is present I am creating the agreement as the delegate for another tenant, if it's not present it's ok, it means I am creating it for myself

I would also rephrase the error and instead of `missingDelegationId` I would make it something like `cannotCreateAgreementForDelegatedEservice`, because the issue is not that the delegationId is missing, but that I am trying to create an agreement for an eservice that I delegated to some other org

```suggestion
    const hasDelegated = delegations.some(
      (d) => d.delegatorId === organizationId
    );
```

Reply:
> Shouldn't this check only if the requester is the delegator?
> 
>     * If there is a delegation and I am the delegator, I cannot create the agreement, so this is correct
> 
>     * However, there can be a delegation for which I am the delegate: in that case, if the delegationId is present I am creating the agreement as the delegate for another tenant, if it's not present it's ok, it means I am creating it for myself
> 
> 
> I would also rephrase the error and instead of `missingDelegationId` I would make it something like `cannotCreateAgreementForDelegatedEservice`, because the issue is not that the delegationId is missing, but that I am trying to create an agreement for an eservice that I delegated to some other org

You're right! I forgot of the case where the delegate could create an agreement for itself!
For the error maybe we could still use `operationNotAllowed` that is used in similar cases